File: src/main/java/client/messages/AbstractMessage.java
package client.messages;

import util.packet.PacketWriter;

public abstract class AbstractMessage implements Message {

    @Override
    public abstract MessageType getType();

    @Override
    public void encode(PacketWriter pw) {
        pw.write(getType().getValue());
        encodeData(pw);
    }

    protected abstract void encodeData(PacketWriter pw);
}

File: src/main/java/client/messages/AbstractMessage.kt
package client.messages

import util.packet.PacketWriter

abstract class AbstractMessage : Message {

    abstract override val type: MessageType

    override fun encode(pw: PacketWriter) {
        pw.write(type.value)
        encodeData(pw)
    }

    protected abstract fun encodeData(pw: PacketWriter)
}